### ISSUE TITLE
Make Specialized logic for extracting language-specific strings from Go

### ISSUE THREAD

**Issue by Arker123** [mandiant/flare-floss#779]
This issue aims to enhance the process of extracting language-specific strings from Go code by implementing specialized logic.
```[tasklist]
### Tasks
```

**Comment by mr-tz** [comment]
https://github.com/CarveSystems/gostringsr2/blob/master/gostringsr2/gostringsr2.py

**Comment by williballenthin** [comment]
this seems to be the declaration for `string` in the Go source code:
https://github.com/golang/go/blob/38e2376f35907ebbb98419f1f4b8f28125bf6aaf/src/go/types/builtins.go#L824-L825

```go
	case _String:
		// unsafe.String(ptr *byte, len IntegerType) string
```

**Comment by mr-tz** [comment]
Neat, that sounds like a good approach if the strings are always laid out like this.

**Comment by mr-tz** [comment]
Is there a common pattern you can see for the references to 1. and 2.? We may have to revert to the "identify assembly sequences" algorithm for that. Side bar: it looks like the strings for 1. and 2. are stored in the same region and both with large chunks of strings. So, maybe they're the same?!

I think 3. may be a special case - we should note but not worry about too much for now.

**Comment by mr-tz** [comment]
These patterns will be helpful for the layout you found initially. However, you discovered that this may not always apply.
Ideally, we come with a generic way that always can recover (most of) the strings.

**Comment by mr-tz** [comment]
I hope so. Have you checked out the extraction script I've [linked](https://github.com/mandiant/flare-floss/issues/779#issuecomment-1576602793) above or any other existing work?

**Comment by mr-tz** [comment]
Ok, I'd suggest we start with the entries that store `<string offset>` and `<string length>`.
- How can we identify where this data starts?
  - is there metadata?
  - can we rely on pefile only or do we need disassembly for this?

Then we can parse this data and show the strings with their correct lengths.

**Comment by mr-tz** [comment]
It looks like either the `.data` or `.rdata` section contain the offset/length information. Initially, we could brute-force enumerate those?

**Comment by mr-tz** [comment]
All great points!
I think strings from the `.text` section and testing across Go versions should be next.

To make this measurable a script to show string coverage would be great. It should show how much of the binaries static strings (like from `string.exe`) we extract with our specialized logic. Then we know we extract X% of strings using method 1, Y% of strings using method 2, etc. And this should highlight blind spots or imperfect algorithms.

What do you think about this?

**Comment by mr-tz** [comment]
> My idea involves examining all operands and specifically focusing on the second operand to determine if it qualifies as a pure string, consisting of alphanumeric and special characters. I believe this approach has the potential to capture strings that may have been overlooked using the conventional breakdown method. I would greatly appreciate your perspective on this idea.

That sounds like a good idea to play around with.

> I wanted to seek your guidance on whether we should test all versions or target a subset of versions, considering the vast array of releases.

Let's start with the most common versions (if we can insight into prevalence) and hope that a generic approach works for most versions. If not, we can handle special cases later on.

**Comment by williballenthin** [comment]
perhaps we could use docker to specify the test executable compilation. this would make it easy to swap the compiler versions, assuming our source syntax doesn’t have to change (i don’t think it should). building using Go 1.0-1.20 would be a reasonable set of compiler versions. 

i would not recommend doing this by hand; it should be mostly/fully scriptable with a little work. let me know if i can lend a hand.

**Comment by mr-tz** [comment]
For visualization something like https://github.com/moreati/python-pixd would be neat to get an idea of which ranges we're covering on high level. It could first indicate all (potential) strings and then lay our findings over that.

![floss-vis](https://github.com/mandiant/flare-floss/assets/17606537/99acb635-0767-4a7e-afe5-a83b92460b64)

**Comment by mr-tz** [comment]
Maybe we can use https://github.com/wader/fq or a similar tool as well (pass in the binary and found string offsets from our extractor).

**Comment by williballenthin** [comment]
great work investigating this!

@mr-tz shall we add all these test files to the repo? checking before we commit 60mb+ (guessing) to the repo. i think LFS is already enabled.

i think it would be really neat to have test cases that show exactly the coverage of our algorithms, proving they work against all versions (eventually).

**Comment by williballenthin** [comment]
btw @Arker123 explaining the string layout across versions and its evolution is a solid blog post. not asking you to do more work but it's an interesting topic and you're about to become an expert.

**Comment by mr-tz** [comment]
> shall we add all these test files to the repo? checking before we commit 60mb+ (guessing) to the repo. i think LFS is already enabled.

I think so. Maybe we can store them zipped (if that helps with size).

**Comment by mr-tz** [comment]
To summarize, we've identified:
- strings stored in data sections in form of `<pointer>, <size>` (handled)
- shorter strings stored in .text section as inline constants (partly handled, see #805)
- strings stored in data sections and referenced from .text section (currently not handled)

The last type is what we want to prioritize next. The second type is of lower priority.

**Comment by mr-tz** [comment]
These look like the function names from the file metadata and I think they can be helpful. Even better when we show that clearly to the user (and break out other identified chunks like user-defined strings, etc.).

There is tools that deal with this (e.g. GoReSym https://www.mandiant.com/resources/blog/golang-internals-symbol-recovery) so we can mark it lower priority but also inspect the existing [source](https://github.com/mandiant/GoReSym).

**Comment by mr-tz** [comment]
Can you explain what the columns in the log represent?
We should handle missing import directories gracefully.

**Comment by mr-tz** [comment]
closed via #792
